File: android/app/src/main/kotlin/com/vortx/android/player/TrackSelector.kt
package com.vortx.android.player

import com.vortx.android.model.TrackPreferences

/**
 * Picks the audio and subtitle track to auto-select from the available tracks and the user's
 * preferences. Pure and side-effect free (so it is unit-testable), and engine-agnostic: it operates on
 * the chrome's own [PlayerTrack] list, so ONE selector serves both libmpv and ExoPlayer.
 *
 * The Android port of Apple `app/Sources/Player/TrackSelector.swift`, kept faithful to its policy so both
 * platforms auto-pick the same track for the same file + preferences (issue #76: a Turkish-only
 * preference must not silently play a French dub off a multi-language European release). [PlayerTrack]
 * carries the same fields Apple's `MPVTrack` did (id / lang / title / forced), so the selection rules map
 * across one-for-one.
 */
object TrackSelector {

    /**
     * The audio and subtitle track ids to select. A [subtitleId] of `-1` means "subtitles off"; a `null`
     * [audioId] means "leave the engine's default" (neither the user's chain nor the English fallback
     * matched, so the container/first-track default the engine already chose is kept).
     */
    data class Selection(val audioId: Int?, val subtitleId: Int?)

    /**
     * The audio and subtitle track to auto-select. Mirrors Apple `TrackSelector.select`.
     *
     * No track matching the user's language chain (#76, ozdek's report) does NOT leave the pick to the
     * engine default: unpicked, both engines defer to the container's default/first audio track, which on
     * multi-language European releases is frequently the local dub, so a Turkish-only preference played
     * French. It falls back to English, the original language of the overwhelming majority of the catalog.
     * The subtitle policy keys off the CHAIN match, not the fallback: English-fallback audio counts as
     * foreign-language content, so full subtitles in the user's language still auto-enable.
     */
    fun select(
        audio: List<PlayerTrack>,
        subtitles: List<PlayerTrack>,
        preferences: TrackPreferences,
    ): Selection {
        val chainPick = firstMatch(audio, preferences.audioLanguages, preferences.rejectTerms)
        val audioPick = chainPick ?: firstMatch(audio, listOf("en"), preferences.rejectTerms)
        val subtitle = selectSubtitle(subtitles, preferences, gotPreferredAudio = chainPick != null)
        return Selection(audioPick?.id, subtitle)
    }

    /**
     * Whether the chrome should fall back to an EXTERNAL (add-on) subtitle for this load: the user's
     * preferences wanted FULL subtitles but no embedded track matched the language chain. Mirrors
     * [selectSubtitle]'s policy exactly, so the add-on fallback never fires when the user asked for
     * subtitles off / forced-only while their preferred audio is present, and never fires when an embedded
     * track already satisfies the chain (the embedded auto-select handles that case).
     */
    fun wantsExternalSubtitle(
        audio: List<PlayerTrack>,
        subtitles: List<PlayerTrack>,
        preferences: TrackPreferences,
    ): Boolean {
        val gotPreferredAudio = firstMatch(audio, preferences.audioLanguages, preferences.rejectTerms) != null
        // With preferred audio present, only the "always" policy shows full subtitles; off/forced-only must
        // not trigger a full external sub. Foreign-language content (no preferred audio) always wants them.
        if (gotPreferredAudio && preferences.forcedPolicy != TrackPreferences.ForcedPolicy.ALWAYS) return false
        return firstMatch(subtitles, preferences.subtitleLanguages, preferences.rejectTerms) == null
    }

    /** First track whose language matches the priority list and whose title isn't rejected. */
    private fun firstMatch(tracks: List<PlayerTrack>, languages: List<String>, reject: List<String>): PlayerTrack? {
        for (lang in languages) {
            tracks.firstOrNull { matches(it.lang, lang) && !isRejected(it, reject) }?.let { return it }
        }
        return null
    }

    /** The subtitle track id to select, or `-1` for "off". Mirrors Apple `selectSubtitle`. */
    private fun selectSubtitle(
        subs: List<PlayerTrack>,
        preferences: TrackPreferences,
        gotPreferredAudio: Boolean,
    ): Int {
        if (subs.isEmpty()) return SUBTITLE_OFF
        // Foreign-language content (no preferred audio matched): show full subtitles so you can follow.
        if (!gotPreferredAudio) {
            return firstMatch(subs, preferences.subtitleLanguages, preferences.rejectTerms)?.id ?: SUBTITLE_OFF
        }
        return when (preferences.forcedPolicy) {
            TrackPreferences.ForcedPolicy.OFF -> SUBTITLE_OFF
            TrackPreferences.ForcedPolicy.ALWAYS ->
                firstMatch(subs, preferences.subtitleLanguages, preferences.rejectTerms)?.id ?: SUBTITLE_OFF
            TrackPreferences.ForcedPolicy.FORCED -> selectForcedSubtitle(subs, preferences)
        }
    }

    /**
     * Match by the container's FORCED disposition flag FIRST: real forced tracks are flagged
     * (AV_DISPOSITION_FORCED / mpv track-list `forced` / ExoPlayer SELECTION_FLAG_FORCED), not labelled
     * "forced" in the title, so the old title-only match never fired for them. Prefer a forced track in a
     * preferred subtitle language, then ANY forced track (forced subs are meant to show regardless of
     * language), then fall back to the legacy title-contains-"forced" heuristic for the rare container that
     * only labels forced in its title. Off if nothing qualifies. Mirrors Apple's `.forced` branch.
     */
    private fun selectForcedSubtitle(subs: List<PlayerTrack>, preferences: TrackPreferences): Int {
        for (lang in preferences.subtitleLanguages) {
            subs.firstOrNull { it.forced && matches(it.lang, lang) && !isRejected(it, preferences.rejectTerms) }
                ?.let { return it.id }
        }
        subs.firstOrNull { it.forced && !isRejected(it, preferences.rejectTerms) }?.let { return it.id }
        for (lang in preferences.subtitleLanguages) {
            subs.firstOrNull {
                matches(it.lang, lang) &&
                    it.title.lowercase().contains("forced") &&
                    !isRejected(it, preferences.rejectTerms)
            }?.let { return it.id }
        }
        return SUBTITLE_OFF
    }

    private fun isRejected(track: PlayerTrack, reject: List<String>): Boolean {
        val title = track.title.lowercase()
        return reject.any { it.isNotEmpty() && title.contains(it.lowercase()) }
    }

    /** Language match, tolerant of 2- vs 3-letter ISO codes (en/eng) and region suffixes (en-US). */
    fun matches(a: String?, b: String?): Boolean {
        val ca = canonical(a)
        return ca.isNotEmpty() && ca == canonical(b)
    }

    /** Reduce a language code to a canonical 2-letter form (eng -> en, en-US -> en, ja -> ja). */
    fun canonical(code: String?): String {
        val base = code?.lowercase()?.substringBefore('-') ?: ""
        if (base.length == 3) alpha3to2[base]?.let { return it }
        return base.take(2)
    }

    /** Subtitle sentinel: no track selected (subtitles off). */
    const val SUBTITLE_OFF = -1

    /**
     * 3-letter codes whose 2-letter form is NOT their first two letters, in both ISO 639-2/T and /B
     * spellings (Matroska muxers write the B codes: "rum", "slo", "per", ...), plus the legacy
     * OpenSubtitles codes add-ons still send ("pob" = Brazilian Portuguese, "scc"/"scr" = Serbian/
     * Croatian). Without an entry the take(2) fallback can cross languages entirely: "est" (Estonian) would
     * match an "es" (Spanish) preference and "rum" (Romanian) a "ru" (Russian) one. Byte-for-byte the Apple
     * `alpha3to2` table so both platforms canonicalize identically.
     */
    private val alpha3to2: Map<String, String> = mapOf(
        "eng" to "en", "spa" to "es", "fra" to "fr", "fre" to "fr", "deu" to "de", "ger" to "de",
        "ita" to "it", "por" to "pt", "rus" to "ru", "jpn" to "ja", "kor" to "ko", "zho" to "zh",
        "chi" to "zh", "ara" to "ar", "hin" to "hi", "nld" to "nl", "dut" to "nl", "swe" to "sv",
        "nor" to "no", "dan" to "da", "fin" to "fi", "pol" to "pl", "tur" to "tr", "tha" to "th",
        "vie" to "vi", "ind" to "id", "heb" to "he", "ell" to "el", "gre" to "el", "ces" to "cs", "cze" to "cs",
        "ron" to "ro", "rum" to "ro", "bul" to "bg", "slk" to "sk", "slo" to "sk", "fas" to "fa",
        "per" to "fa", "est" to "et", "lav" to "lv", "lit" to "lt", "isl" to "is", "ice" to "is",
        "mkd" to "mk", "mac" to "mk", "sqi" to "sq", "alb" to "sq", "hye" to "hy", "arm" to "hy",
        "kat" to "ka", "geo" to "ka", "eus" to "eu", "baq" to "eu", "cym" to "cy", "wel" to "cy",
        "msa" to "ms", "may" to "ms", "ben" to "bn", "mal" to "ml", "mar" to "mr", "kan" to "kn",
        "mya" to "my", "bur" to "my", "khm" to "km", "lao" to "lo", "kaz" to "kk", "bos" to "bs",
        "mlt" to "mt", "gle" to "ga", "fil" to "tl", "tgl" to "tl", "pob" to "pt", "scc" to "sr", "scr" to "hr",
    )
}

File: app/Sources/Player/TrackSelector.swift
import Foundation

/// Picks the audio and subtitle track to auto-select from the available tracks and the user's
/// preferences. Pure and side-effect free, so it is unit-testable. Shared by both players.
enum TrackSelector {
    /// Selection setters may have side effects even when the requested row is already active (MPV arms a
    /// cache hold; AVPlayer's source-audio lane can remount). Make that no-op executable for both engines.
    static func shouldApplyAudioSelection(_ id: Int, to tracks: [MPVTrack]) -> Bool {
        !tracks.contains { $0.id == id && $0.selected }
    }

    /// A remux has already made its primary audio choice before AVPlayer can load its track inventory.
    /// Suppress only the automatic setter in that case; explicit picker actions call the engine directly.
    static func automaticAudioSelection(_ id: Int?,
                                        remuxOwnsInitialSelection: Bool) -> Int? {
        remuxOwnsInitialSelection ? nil : id
    }

    /// The audio and subtitle track ids to select. A subtitle id of -1 means "off"; a nil audio id
    /// means "leave the engine's default" (neither the user's chain nor the English fallback matched).
    static func select(audio: [MPVTrack], subtitles: [MPVTrack], preferences p: TrackPreferences) -> (audio: Int?, subtitle: Int?) {
        let chainPick = firstAudioMatch(audio, languages: p.audioLanguages, reject: p.rejectTerms)
        // No track matches the user's language chain (#76, ozdek's report): do NOT leave the pick to the
        // engine default. Unpicked, both engines defer to the container's default/first audio track, which
        // on multi-language European releases is frequently the local dub, so a Turkish-only preference
        // played French. Fall back to English, the original language of the overwhelming majority of the
        // catalog (the app has no per-title original-language metadata to prefer first, and MPVTrack
        // carries no channel counts to break ties on, so English is the strongest fallback available).
        // A file with neither a chain match nor an English track still leaves the engine default.
        // The subtitle policy keys off the CHAIN match, not the fallback: English-fallback audio counts as
        // foreign-language content, so full subtitles in the user's language still auto-enable.
        let audioPick = chainPick ?? firstAudioMatch(audio, languages: ["en"], reject: p.rejectTerms)
        let subtitle = selectSubtitle(subtitles, preferences: p, gotPreferredAudio: chainPick != nil)
        return (audioPick?.id, subtitle)
    }

    /// Whether the chrome should fall back to an EXTERNAL (add-on) subtitle for this load: the user's
    /// preferences wanted FULL subtitles but no embedded track matched the language chain. Mirrors
    /// `selectSubtitle`'s policy logic exactly, so the add-on fallback never fires when the user asked for
    /// subtitles off / forced-only while their preferred audio is present, and never fires when an embedded
    /// track already satisfies the chain (the embedded auto-select handles that case).
    static func wantsExternalSubtitle(audio: [MPVTrack], subtitles: [MPVTrack], preferences p: TrackPreferences) -> Bool {
        let gotPreferredAudio = firstMatch(audio, languages: p.audioLanguages, reject: p.rejectTerms) != nil
        // With preferred audio present, only the "always" policy shows full subtitles; off/forced-only must
        // not trigger a full external sub. Foreign-language content (no preferred audio) always wants them.
        if gotPreferredAudio && p.forcedPolicy != .always { return false }
        return firstMatch(subtitles, languages: p.subtitleLanguages, reject: p.rejectTerms) == nil
    }

    /// Audio rows within one language tier are otherwise ordered only by the engine/container. Keep the row
    /// already playing so an automatic pass cannot turn a no-op into a source-audio remount.
    private static func firstAudioMatch(_ tracks: [MPVTrack], languages: [String], reject: [String]) -> MPVTrack? {
        for lang in languages {
            var firstEligible: MPVTrack?
            for track in tracks where track.isSelectable
                && matches(track.lang, lang)
                && !isRejected(track, reject) {
                if track.selected { return track }
                if firstEligible == nil { firstEligible = track }
            }
            if let firstEligible { return firstEligible }
        }
        return nil
    }

    /// First track whose language matches the priority list and whose title isn't rejected.
    private static func firstMatch(_ tracks: [MPVTrack], languages: [String], reject: [String]) -> MPVTrack? {
        for lang in languages {
            if let t = tracks.first(where: {
                $0.isSelectable
                    && matches($0.lang, lang)
                    && !isRejected($0, reject)
            }) { return t }
        }
        return nil
    }

    private static func selectSubtitle(_ subs: [MPVTrack], preferences p: TrackPreferences, gotPreferredAudio: Bool) -> Int? {
        let selectable = subs.filter(\.isSelectable)
        guard !selectable.isEmpty else { return -1 }
        // Foreign-language content (no preferred audio matched): show full subtitles so you can follow.
        if !gotPreferredAudio {
            return firstMatch(selectable, languages: p.subtitleLanguages, reject: p.rejectTerms)?.id ?? -1
        }
        switch p.forcedPolicy {
        case .off:
            return -1
        case .always:
            return firstMatch(selectable, languages: p.subtitleLanguages, reject: p.rejectTerms)?.id ?? -1
        case .forced:
            // Match by the container's FORCED disposition flag FIRST: real forced tracks are flagged
            // (AV_DISPOSITION_FORCED / mpv track-list forced), not labelled "forced" in the title, so the old
            // title-only match never fired for them and forced subtitles never turned on. Prefer a forced track
            // in a preferred subtitle language, then ANY forced track (forced subs are meant to show regardless
            // of language), then fall back to the legacy title-contains-"forced" heuristic for the rare
            // container that only labels forced in its title. Off if nothing qualifies.
            for lang in p.subtitleLanguages {
                if let t = selectable.first(where: { $0.forced && matches($0.lang, lang) && !isRejected($0, p.rejectTerms) }) {
                    return t.id
                }
            }
            if let t = selectable.first(where: { $0.forced && !isRejected($0, p.rejectTerms) }) {
                return t.id
            }
            for lang in p.subtitleLanguages {
                if let t = selectable.first(where: { matches($0.lang, lang) && $0.title.lowercased().contains("forced") && !isRejected($0, p.rejectTerms) }) {
                    return t.id
                }
            }
            return -1
        }
    }

    private static func isRejected(_ track: MPVTrack, _ reject: [String]) -> Bool {
        let title = track.title.lowercased()
        return reject.contains { !$0.isEmpty && title.contains($0.lowercased()) }
    }

    /// Language match, tolerant of 2- vs 3-letter ISO codes (en/eng) and region suffixes (en-US).
    static func matches(_ a: String, _ b: String) -> Bool {
        let ca = canonical(a)
        return !ca.isEmpty && ca == canonical(b)
    }

    /// Filter an external-subtitle list to the preferred subtitle languages when the opt-in
    /// `TrackPreferences.subtitlesOnlyPreferred` toggle is on; a no-op (returns `items` unchanged) otherwise.
    /// Unknown/undetermined languages ("", "und", "unknown") are ALWAYS kept so the filter can never hide every
    /// option (add-on and embedded-upload subs frequently report no language). Uses the tolerant `matches`, so
    /// "tur"/"tr-TR" still satisfy a "tr" preference. Generic over the row type via a `language` accessor so
    /// both the add-on rows (`AddonSubtitle`) and the community-pooled rows share one path.
    static func keepingPreferredSubtitleLanguages<T>(_ items: [T], language: (T) -> String) -> [T] {
        guard TrackPreferences.subtitlesOnlyPreferred else { return items }
        let prefs = TrackPreferences.current.subtitleLanguages
        guard !prefs.isEmpty else { return items }
        return items.filter { item in
            let raw = language(item).trimmingCharacters(in: .whitespaces).lowercased()
            if raw.isEmpty || raw == "und" || raw == "unknown" { return true }
            return prefs.contains { matches(language(item), $0) }
        }
    }

    /// Reduce a language code to its canonical base identity (eng → en, en-US → en, ja → ja).
    static func canonical(_ code: String) -> String {
        AudioLanguagePolicy.canonical(code)
    }
}
